File: test/relation.cpp
#include "sys-sage.hpp"
#include <boost/ut.hpp>
#include <iostream>
#include <vector>

namespace ut = boost::ut;
using namespace sys_sage;

// Disable std::cerr to prevent logging in `GetComponent()` when causing an
// exception. We need RAII for that to make sure that std::cerr gets enabled again.
//
// This wrapper assumes that the stream is valid during the lifetime of the wrapper
class stream_suppressor {
public:
  stream_suppressor(std::ostream &_stream) : stream(_stream), buf (_stream.rdbuf(nullptr)) {}
  ~stream_suppressor() { stream.rdbuf(buf); }

private:
  std::ostream &stream;
  std::streambuf *buf;
};

ut::suite<"relation"> _ = []
{
  ut::test("interconnection") = []
  {
    Component foo, bar;
    std::vector<Component *> v {&foo, &bar};
    Relation r (v);

    ut::expect(ut::that % r.GetComponent(0) == &foo);
    ut::expect(ut::that % r.GetComponent(1) == &bar);
    {
      stream_suppressor suppressor (std::cerr);
      ut::expect(ut::that % r.GetComponent(2) == nullptr);
    }
    ut::expect(ut::that % r.GetComponents() == v);

    ut::expect(ut::that % foo.GetRelations(RelationType::Relation) == std::vector<Relation *> {&r});
    ut::expect(ut::that % bar.GetRelations(RelationType::Relation) == std::vector<Relation *> {&r});
  };

  ut::test("deletion") = []
  {
    Component foo, bar;
    std::vector<Component *> v {&foo, &bar};
    Relation *r = new Relation (v);
    // NOTE: this only works if r doesn't live on the stack
    r->Delete();

    ut::expect(ut::that % foo.GetRelations(RelationType::Relation).size() == 0U);
    ut::expect(ut::that % bar.GetRelations(RelationType::Relation).size() == 0U);
  };

  ut::test("getters & setters") = []
  {
    std::vector<Component *> v;
    Relation r (v);

    ut::expect(ut::that % r.GetType() == RelationType::Relation);

    r.SetId(2);
    ut::expect(ut::that % r.GetId() == 2);
    ut::expect(ut::that % r.IsOrdered());
  };

  ut::test("adding & updating components") = []
  {
    std::vector<Component *> v;
    Relation r (v);
    ut::expect(ut::that % r.GetComponents().size() == 0U);

    Component foo;
    r.AddComponent(&foo);
    ut::expect(ut::that % r.ContainsComponent(&foo));

    Component bar;
    {
      stream_suppressor suppressor (std::cerr);
      ut::expect(ut::that % r.UpdateComponent(&bar, &bar) == 1);
    }

    r.UpdateComponent(&foo, &bar);
    ut::expect(ut::that % r.GetComponents() == std::vector<Component *> {&bar});

    Component foobar;
    {
      stream_suppressor suppressor (std::cerr);
      ut::expect(ut::that % r.UpdateComponent(1, &foobar) == 1);
    }

    r.UpdateComponent(0, &foobar);
    ut::expect(ut::that % r.GetComponents() == std::vector<Component *> {&foobar});
  };

  ut::test("attributes") = []
  {
    std::vector<Component *> v (0);
    Relation r (v);

    int *i = new int (3);
    r.attrib["foo"] = reinterpret_cast<void *>(i);
    auto it = r.attrib.find("foo");
    ut::expect(ut::that % (it != r.attrib.end()));
    ut::expect(ut::that % *reinterpret_cast<int *>(it->second) == 3);
    delete i;
  };

  ut::test("inheritance") = []
  {
    ut::test("DataPath") = []
    {
      Component foo, bar;
      Relation *r = new DataPath(&foo, &bar, DataPathOrientation::Oriented, DataPathType::Any);
      ut::expect(ut::that % r->GetType() == RelationType::DataPath);
      dynamic_cast<DataPath *>(r)->SetBandwidth(1.0);
      dynamic_cast<DataPath *>(r)->SetLatency(2.0);
      r->Delete();
    };

    ut::test("QuantumGate") = []
    {
      Qubit foo, bar;
      std::vector<Qubit *> v {&foo, &bar};
      // why do I need to pass in the number of qubits (i.e. v.size()) and also v?
      // can't the number of qubits be inferred by just passing in the vector of qubits?
      Relation *r = new QuantumGate(v.size(), v);

      ut::expect(ut::that % r->GetType() == RelationType::QuantumGate);
      dynamic_cast<QuantumGate *>(r)->SetGateProperties("cx", 1.0, "[1 0 0 0; 0 1 0 0; 0 0 0 1; 0 0 1 0]");
      ut::expect(ut::that % dynamic_cast<QuantumGate *>(r)->GetQuantumGateType() == QuantumGateType::Cnot);
      ut::expect(ut::that % dynamic_cast<QuantumGate *>(r)->GetFidelity() == 1.0);
      ut::expect(ut::that % (dynamic_cast<QuantumGate *>(r)->GetUnitary() == "[1 0 0 0; 0 1 0 0; 0 0 0 1; 0 0 1 0]"));
      r->Delete();
    };

    ut::test("CouplingMap") = []
    {
      Qubit foo, bar;
      Relation *r = new CouplingMap(&foo, &bar);
      ut::expect(ut::that % r->GetType() == RelationType::CouplingMap);
      dynamic_cast<CouplingMap *>(r)->SetFidelity(1.0);
      ut::expect(ut::that % dynamic_cast<CouplingMap *>(r)->GetFidelity() == 1.0);
      r->Delete();
    };
  };
};

File: test/relation.py
import py_sys_sage as pysage
import unittest
import sys
import os

#class stream_suppressor:
#    def __init__(self, stream):
#        self.stream = stream
#
#    def __enter__(self):
#        self.original_stream = getattr(sys, self.stream)
#        self.devnull = open(os.devnull, "w")
#        setattr(sys, self.stream, self.devnull)
#
#    def __exit__(self, *_):
#        setattr(sys, self.stream, self.original_stream)
#        self.devnull.close()

class TestRelations(unittest.TestCase):
    def test_interconnection(self):
        foo = pysage.Component()
        bar = pysage.Component()
        v = [foo, bar]
        r = pysage.Relation(v)
        self.assertEqual(r.GetComponent(0), foo)
        self.assertEqual(r.GetComponent(1), bar)
        #with stream_suppressor("stderr"):
        #    self.assertEqual(r.GetComponent(2), None)
        self.assertEqual(r.GetComponent(2), None)
        self.assertEqual(r.components, v)
        self.assertEqual(foo.GetRelations(pysage.RELATION_TYPE_RELATION), [r])
        self.assertEqual(bar.GetRelations(pysage.RELATION_TYPE_RELATION), [r])

    def test_removal(self):
        foo = pysage.Component()
        bar = pysage.Component()
        v = [foo, bar]
        r = pysage.Relation(v)
        r.RemoveComponent(foo)
        r.RemoveComponent(bar)
        
        self.assertEqual(len(foo.GetRelations(pysage.RELATION_TYPE_RELATION)), 0)
        self.assertEqual(len(bar.GetRelations(pysage.RELATION_TYPE_RELATION)), 0)

    def test_getters_and_setters(self):
        r = pysage.Relation([])

        self.assertEqual(r.type, pysage.RELATION_TYPE_RELATION)

        r.id = 2
        self.assertEqual(r.id, 2)
        self.assertTrue(r.ordered)

    def test_adding_and_updating_components(self):
        r = pysage.Relation([])
        self.assertEqual(len(r.components), 0)

        foo = pysage.Component()
        r.AddComponent(foo)
        self.assertTrue(r.ContainsComponent(foo))

        bar = pysage.Component()
        #with stream_suppressor("stderr"):
        #    self.assertEqual(r.UpdateComponent(bar, bar), 1)
        self.assertEqual(r.UpdateComponent(bar, bar), 1)

        r.UpdateComponent(foo, bar)
        self.assertEqual(r.components, [bar])

        foobar = pysage.Component()
        #with stream_suppressor("stderr"):
        #    self.assertEqual(r.UpdateComponent(1, foobar), 1)
        self.assertEqual(r.UpdateComponent(1, foobar), 1)

        r.UpdateComponent(0, foobar)
        self.assertEqual(r.components, [foobar])

    def test_attributes(self):
        r = pysage.Relation([])

        r.SetAttribute("foo", 1)
        r.SetAttribute("bar", 2.0)
        r.SetAttribute("foobar", "test")

        self.assertEqual(r.GetAttribute("foo"), 1)
        self.assertEqual(r.GetAttribute("bar"), 2.0)
        self.assertEqual(r.GetAttribute("foobar"), "test")

        r.EraseAttribute("foo")
        self.assertEqual(r.GetAttribute("foo"), None)

        r.ClearAttributes()
        self.assertEqual(r.GetAttributesSize(), 0)

    def test_inheritance_data_path(self):
        foo = pysage.Component()
        bar = pysage.Component()

        r = pysage.DataPath(foo, bar, pysage.DATAPATH_ORIENTATION_ORIENTED, pysage.DATAPATH_CATEGORY_ANY)
        # test if inherited class can access members of the base class
        self.assertEqual(r.type, pysage.RELATION_TYPE_DATAPATH)

    def test_inheritance_quantum_gate(self):
        foo = pysage.Qubit()
        bar = pysage.Qubit()
        v = [foo, bar]
        r = pysage.QuantumGate(len(v), v)

        self.assertEqual(r.type, pysage.RELATION_TYPE_QUANTUMGATE)

        r.SetGateProperties("cx", 1.0, "[1 0 0 0; 0 1 0 0; 0 0 0 1; 0 0 1 0]")
        self.assertEqual(r.gate_type, pysage.QUANTUMGATE_CATEGORY_CNOT)
        self.assertEqual(r.fidelity, 1.0)
        self.assertEqual(r.unitary, "[1 0 0 0; 0 1 0 0; 0 0 0 1; 0 0 1 0]")

    def test_inheritance_coupling_map(self):
        foo = pysage.Qubit()
        bar = pysage.Qubit()
        r = pysage.CouplingMap(foo, bar)

        self.assertEqual(r.type, pysage.RELATION_TYPE_COUPLINGMAP)

        r.fidelity = 1.0
        self.assertEqual(r.fidelity, 1.0)

if __name__ == "__main__":
    unittest.main()
